frb_rust/src/ffi.rs

```diff
@@ -0,0 +1,214 @@
+//! Stubs for allo_isolate on the web.
+
+#[cfg(not(target_family = "wasm"))]
+pub use allo_isolate::{ffi::DartCObject, *};
+#[cfg(not(target_family = "wasm"))]
+pub type MessagePort = i64;
+
+#[macro_export]
+macro_rules! console_log {
+    ($lit:literal) => {
+        #[allow(unused_unsafe)]
+        unsafe { $crate::ffi::log($lit) }
+    };
+    ($( $tt:tt )*) => {
+        #[allow(unused_unsafe)]
+        unsafe { $crate::ffi::log(&format!($( $tt )*)) }
+    };
+}
+
+#[macro_export]
+macro_rules! console_error {
+    ($lit:literal) => {
+        #[allow(unused_unsafe)]
+        unsafe { $crate::ffi::error($lit) }
+    };
+    ($( $tt:tt )*) => {
+        #[allow(unused_unsafe)]
+        unsafe { $crate::ffi::error(&format!($( $tt )*)) }
+    };
+}
+
+#[cfg(target_family = "wasm")]
+pub use web::*;
+#[cfg(target_family = "wasm")]
+mod web {
+    use js_sys::{global, Array};
+    use std::iter::FromIterator;
+    pub use wasm_bindgen::closure::Closure;
+    use wasm_bindgen::prelude::*;
+    use wasm_bindgen::JsCast;
+    use web_sys::{DedicatedWorkerGlobalScope, Worker};
+    pub type MessagePort = web_sys::MessagePort;
+
+    pub type DartCObject = JsValue;
+
+    pub trait IntoDart {
+        fn into_dart(self) -> DartCObject;
+    }
+
+    pub trait IntoDartExceptPrimitive: IntoDart {}
+    impl IntoDartExceptPrimitive for JsValue {}
+
+    impl IntoDart for () {
+        #[inline]
+        fn into_dart(self) -> DartCObject {
+            JsValue::undefined()
+        }
+    }
+    macro_rules! delegate {
+        ($( $ty:ty )*) => {$(
+            impl IntoDart for $ty {
+                #[inline]
+                fn into_dart(self) -> DartCObject {
+                    DartCObject::from(self)
+                }
+            }
+        )*};
+    }
+    macro_rules! delegate_buffer {
+        ($( $ty:ty => $buffer:ty )*) => {$(
+            impl IntoDart for $ty {
+                #[inline]
+                fn into_dart(self) -> DartCObject {
+                    <$buffer>::from(self.as_slice()).into()
+                }
+            }
+        )*};
+    }
+    // Orphan rules do not benefit our implementation, so we manually delegate From conversions here.
+    delegate! {
+        bool i8 u8 i16 u16 i32 u32 i64 u64 i128 u128 isize usize
+        f32 f64 &str String JsValue
+    }
+    delegate_buffer! {
+        Vec<i8> => js_sys::Int8Array
+        Vec<u8> => js_sys::Uint8Array
+        Vec<i16> => js_sys::Int16Array
+        Vec<u16> => js_sys::Uint16Array
+        Vec<i32> => js_sys::Int32Array
+        Vec<u32> => js_sys::Uint32Array
+        Vec<f32> => js_sys::Float32Array
+        Vec<f64> => js_sys::Float64Array
+        ZeroCopyBuffer<Vec<i8>> => js_sys::Int8Array
```

Comment:
To have ZeroCopyBuffer be actually "zero-copy", we would need to pass it as a transferrable in a `postMessage`, which would couple it with the implementation of the WASM worker pool. This might not be desirable especially for merging upstream.

Reply:
Agree. So maybe we can currently document ZeroCopyBuffer as unsupported in the Web, and make it fallback to normal vectors.